packages/monitor-v2/src/monitor-og/SnapshotVerification.ts

```diff
@@ -0,0 +1,342 @@
+import fetch from "node-fetch";
+import { request } from "graphql-request";
+import { gql } from "graphql-tag";
+
+import { ethersUtils, MonitoringParams, tryHexToUtf8String } from "./common";
+
+// If there are multiple transactions within a batch, they are aggregated as multiSend in the mainTransaction.
+interface MainTransaction {
+  to: string;
+  data: string;
+  value: string;
+  operation: string;
+}
+
+// We only include properties that will be verified by the bot.
+interface SafeSnapSafe {
+  txs: { mainTransaction: MainTransaction }[];
+  network: string;
+  umaAddress: string;
+}
+
+interface SafeSnapPlugin {
+  safeSnap: { safes: SafeSnapSafe[] };
+}
+
+// We only type the properties required in the verification. IPFS hosted proposal object includes additional properties
+// that we don't look at.
+interface SnapshotProposalIpfs {
+  space: string;
+  type: string;
+  choices: string[];
+  start: number;
+  end: number;
+  plugins: Partial<SafeSnapPlugin>;
+}
+
+// We only type the properties requested in the GraphQL query. This extends SnapshotProposalIpfs, but we need to
+// override the space property.
+interface SnapshotProposalGraphql extends Omit<SnapshotProposalIpfs, "space"> {
+  state: string;
+  space: { id: string };
+  scores: number[];
+  quorum: number;
+  scores_total: number;
+}
+
+interface GraphqlData {
+  proposals: SnapshotProposalGraphql[];
+}
+
+interface IpfsData {
+  data: {
+    message: SnapshotProposalIpfs;
+  };
+}
+
+export interface VerificationResponse {
+  verified: boolean;
+  error?: string;
+}
+
+interface RulesParameters {
+  space: string;
+  quorum: number;
+  votingPeriod: number;
+}
+
+// Returns null if the rules string does not match the expected template.
+const parseRules = (rules: string): RulesParameters | null => {
+  // This is based on the template from Zodiac app at
+  // https://github.com/gnosis/zodiac-safe-app/blob/79dbb72af506f60fcc16599516ce48f893393b29/packages/app/src/views/AddModule/wizards/OptimisticGovernorModule/OptimisticGovernorModuleModal.tsx#L136
+  const regex = /^I assert that this transaction proposal is valid according to the following rules: Proposals approved on Snapshot, as verified at https:\/\/snapshot\.org\/#\/(.*), are valid as long as there is a minimum quorum of (\d+) and a minimum voting period of (\d+) hours and it does not appear that the Snapshot voting system is being exploited or is otherwise unavailable\. The quorum and voting period are minimum requirements for a proposal to be valid\. Quorum and voting period values set for a specific proposal in Snapshot should be used if they are more strict than the rules parameter\. The explanation included with the on-chain proposal must be the unique IPFS identifier for the specific Snapshot proposal that was approved or a unique identifier for a proposal in an alternative voting system approved by DAO social consensus if Snapshot is being exploited or is otherwise unavailable.$/;
```

Comment:
Very minor nit: https:\/\/snapshot\.org\/#\/(.*) could capture a ton of text that has nothing to do with the link, right (as long as the exact text after matches).

Do you think we should constrain the .* a bit more to exclude things like spaces that aren't allowed in urls (and other non-url chars)?

Reply:
Agree, for now we can use more restrictive `([a-zA-Z0-9-.]+)` for space names

Reply:
Sounds good!